sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -311,6 +311,35 @@ public void resolveWritableFields(boolean value) {
         forWrite = value;
     }
 
+
+    private static Symbol castIfNeededOrFail(Symbol symbolToCast, DataType targetType, ExpressionAnalysisContext context) {
+        DataType sourceType = symbolToCast.valueType();
+        if (sourceType.equals(targetType)) {
+            return symbolToCast;
+        }
+        if (sourceType.isConvertableTo(targetType)) {
+            // cast further below doesn't always fail because it might be lazy as it wraps functions/references inside a cast function.
+            // -> Need to check isConvertableTo to fail eagerly if the cast won't work.
+            try {
+                return cast(symbolToCast, targetType, context);
+            } catch (NumberFormatException e) {
```

Comment:
ClassCastException shouldn't happen because isConvertableTo is checked. IllegalArgumentException will bubble which is okay because the error message might be more detailed (e.g.  where bytefield = 129 will fail with byte value out of range)